### PR TITLE
updated Code and example for single video - Closes #28, Closes #35

### LINKED ISSUE #28: Fix single video dataloader
# Context
This might be broken by our latest changes, we might need it for the linear probing implementation. 

# Solution
- [ ] Discussion video LinearProbin implementation with @robertavram-md 
- [ ] If you choose to move forward with linear prob. in deepcoro_clip investigate possible break

### LINKED ISSUE #35: Add unit tests

### PR DESCRIPTION
**Pull request by robertavram-md**
Fixes #28  and #35

This pull request includes several changes to the project structure, configuration files, and model implementations for the `DeepCORO_CLIP` deep learning model. The changes are mainly focused on reorganizing the project directory, updating configuration parameters, and refining model components.

### Project Structure Updates:
* [`.cursorrules`](diffhunk://#diff-315161dd3154081cb28d119355d3dd6de62ddc800ac1a12f8eb903e59047a018L1-R66): Updated the project directory structure to reflect new organization and added guidelines for navigating the codebase.

### Configuration Updates:
* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L25-R25): Modified the virtual environment setup instructions to use `.venv` instead of `.deepcoro`.
* [`config/base_config.yaml`](diffhunk://#diff-ad781b6517d6292f33686c44469e3576f928320a4b4996cd88fb4411f98e351bL2-R21): Updated training parameters such as `epochs`, `num_workers`, `batch_size`, and dataset parameters like `data_filename` and `multi_video`. Added new parameters `gradient_accumulation_steps` and `num_warmup_percent`. [[1]](diffhunk://#diff-ad781b6517d6292f33686c44469e3576f928320a4b4996cd88fb4411f98e351bL2-R21) [[2]](diffhunk://#diff-ad781b6517d6292f33686c44469e3576f928320a4b4996cd88fb4411f98e351bR39-R40) [[3]](diffhunk://#diff-ad781b6517d6292f33686c44469e3576f928320a4b4996cd88fb4411f98e351bL68-R71)
* [`config/sweep_config.yaml`](diffhunk://#diff-0c511a402026cd6a215826fc9fc71152f5842f698b65b1a598f4f802d8f4f204L1-L65): Removed the old sweep configuration file.
* [`config/sweep_config.yaml.bak`](diffhunk://#diff-8b5c6888e18f33e356af63c1a2a73bdc7a44ff4e5a9572c0c8b70d8ff640a8cbL1-R11): Updated sweep configuration parameters, including `lr`, `optimizer`, `scheduler_type`, `factor`, and added new parameters like `num_warmup_percent` and `stride`. [[1]](diffhunk://#diff-8b5c6888e18f33e356af63c1a2a73bdc7a44ff4e5a9572c0c8b70d8ff640a8cbL1-R11) [[2]](diffhunk://#diff-8b5c6888e18f33e356af63c1a2a73bdc7a44ff4e5a9572c0c8b70d8ff640a8cbL23-R86)
* [`config/sweep_config_multi_video.yaml`](diffhunk://#diff-2030db728f2c0c02a6aa681f582daea586d79bb86f3591f76644f96553393e1dL10-R11): Renamed from `config/sweep_config_3.yaml` and updated sweep configuration parameters. [[1]](diffhunk://#diff-2030db728f2c0c02a6aa681f582daea586d79bb86f3591f76644f96553393e1dL10-R11) [[2]](diffhunk://#diff-2030db728f2c0c02a6aa681f582daea586d79bb86f3591f76644f96553393e1dL33-R33) [[3]](diffhunk://#diff-2030db728f2c0c02a6aa681f582daea586d79bb86f3591f76644f96553393e1dL43-R62) [[4]](diffhunk://#diff-2030db728f2c0c02a6aa681f582daea586d79bb86f3591f76644f96553393e1dL78-R81)
* [`config/sweep_config_single_video.yaml`](diffhunk://#diff-736b873bc7ffd63be441a212ebee0c1a04046133a93c971b47bf7d25cb3254b6R1-R101): Added new sweep configuration file for single video training.

### Model Updates:
* [`models/text_encoder.py`](diffhunk://#diff-91e55674b1f986e241a4551f0d015cde06d21869cd02fb8b5a31655131b272d8R53-R55): Added a check to remove the pooler if it exists, updated the `_freeze_partial_bert` method to skip freezing if `freeze_ratio` is 0, and refined the `forward` method to use pooled features. [[1]](diffhunk://#diff-91e55674b1f986e241a4551f0d015cde06d21869cd02fb8b5a31655131b272d8R53-R55) [[2]](diffhunk://#diff-91e55674b1f986e241a4551f0d015cde06d21869cd02fb8b5a31655131b272d8R72-R74) [[3]](diffhunk://#diff-91e55674b1f986e241a4551f0d015cde06d21869cd02fb8b5a31655131b272d8L90-R100)
* [`models/video_encoder.py`](diffhunk://#diff-64ffc94ca828921d2f8597ea995c4401f073b999ea89c9c46c6df8542111d1b1L78-R80): Added a check to ensure `output_dim` is divisible by `num_heads`.

### Debugging and Logging:
* [`projects/contrastive_pretraining.py`](diffhunk://#diff-93a0c567b98ba3a77bda67e076971b4edc32598b30404fff6dc92bcfa503b44dR126): Added a debug print statement for single video training.

### CODE REVIEW

File: projects/contrastive_pretraining.py [hunk 1]
```diff
@@ -123,6 +123,7 @@ def load_train_objs(
             drop_last=False,
         )
     else:
+        print("DEBUG single video")
```
Comment (JacquesDelfrate) [inline projects/contrastive_pretraining.py:126]: remove debug print or add print statment like print('Use Single video dataset')
and above print('Use multivideo dataset')

File: projects/contrastive_pretraining.py [hunk 2]
```diff
@@ -198,15 +199,9 @@ def load_train_objs(
...
-        {
-            'params': video_encoder.module.proj.parameters(),  # Video projection
-            'lr': config.lr,
-            'name': 'video_proj',
-            'weight_decay': config.video_weight_decay
-        },
         {
             'params': text_encoder.parameters(),  # Entire text encoder
-            'lr': 0.000009,  # Lower learning rate for text encoder
+            'lr': 0.00001,  # Lower learning rate for text encoder
```
Comment (JacquesDelfrate) [inline projects/contrastive_pretraining.py:204]: No specific lr for text encoder ?

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import pytest
+import torch
+from utils.config import HeartWiseConfig
+
+@pytest.fixture
+def test_config():
```
Comment (JacquesDelfrate) [inline tests/conftest.py:6]: nice!

File: utils/losses.py [hunk 4]
```diff
@@ -132,4 +132,48 @@ def contrastive_loss_ddp(
     loss_i2t = F.cross_entropy(logits, targets)
     loss_t2i = F.cross_entropy(logits.t(), targets)
     loss = 0.5 * (loss_i2t + loss_t2i)
-    return loss
\ No newline at end of file
+    return loss
+
+class InfoNCELoss(nn.Module):
```
Comment (JacquesDelfrate) [inline utils/losses.py:137]: Should be added to get_loss_fn() to wrapp loss fonction abstraction calling

File: projects/contrastive_pretraining.py (hunk 1 above)
Comment (robertavram-md) [reply projects/contrastive_pretraining.py]: Done

File: projects/contrastive_pretraining.py (hunk 2 above)
Comment (robertavram-md) [reply projects/contrastive_pretraining.py:203]: Right now I find really low LR works best so I hard coded it.

File: tests/conftest.py (hunk 3 above)
Comment (robertavram-md) [reply tests/conftest.py:6]: Ty

File: utils/losses.py (hunk 4 above)
Comment (robertavram-md) [reply utils/losses.py:137]: Not sure what u mean can u do it?
